### PR TITLE
fix[codegen]: fix `make_setter` overlap with internal calls

### PR DESCRIPTION
**Pull request by charles-cooper**
### What I did
fix https://github.com/vyperlang/vyper/issues/3503

### Commit message

```
fix overlap analysis for `make_setter` when the RHS contains an internal
call. this represents a bug which was partially fixed in 1c8349e;
this commit fixes two variants:

- where the RHS contains an internal call, the analysis needs to take
  into account all storage variables touched by the called function.

- where the RHS contains an external call, the analysis needs to assume
  any storage variable could be touched (this is a conservative
  analysis; a more advanced analysis could limit the assumption to
  storage variables touched by reentrant functions)

---------

Co-authored-by: cyberthirst <cyberthirst.eth@gmail.com>
Co-authored-by: trocher <rocher14@live.fr>
```

### Cute Animal Picture

Put a link to a cute animal picture inside the parenthesis-->

### CODE REVIEW

File: vyper/codegen/ir_node.py [hunk 1]
```diff
@@ -464,6 +464,11 @@ def referenced_variables(self):
 
         return ret
 
+    @cached_property
+    def contains_risky_call(self):
```
Comment (fubuloubu) [inline vyper/codegen/ir_node.py:468]: this is because it could call back?

**Review by fubuloubu — APPROVED** [review]

File: vyper/codegen/ir_node.py (hunk 1 above)
Comment (charles-cooper) [reply vyper/codegen/ir_node.py:468]: yea, bc of reentrancy we have to assume all variables are tainted

we could restrict it to only variables which are touchable from (reentrant) external functions, but that requires more analysis.

once venom optimizer gets more sophisticated we can clean this up by *always* writing to temp buffer and then letting memory alias analysis optimize out the intermediate writes if it deems they are safe

File: tests/functional/codegen/features/test_internal_call.py [hunk 2]
```diff
@@ -624,6 +624,74 @@ def bar() -> Bytes[6]:
...
+    self.a = []
+    # it should now be impossible to read any of [11, 12, 13, 14, 15, 16]
+    return 3
+    """
+    c = get_contract(code)
+
+    assert c.foo() == [1, 2, 3, 4]
+
+
+def test_make_setter_external_call(get_contract):
```
Comment (cyberthirst) [inline tests/functional/codegen/features/test_internal_call.py:670]: is this in the right file?

**Comment by cyberthirst** [comment]
recursion missing:
```python
def test_make_setter_external_call2(get_contract):
    # variant of GH #3503
    code = """
interface A:
   def boo(): nonpayable

a: DynArray[uint256, 10]

@external
def foo() -> DynArray[uint256, 10]:
    self.a = [1, 2, self.baz(), 4]
    return self.a # returns [11, 12, 3, 4]


@internal
def baz() -> uint256:
    extcall A(self).boo()
    return 3

@external
def boo():
    self.a = [11, 12, 13, 14, 15, 16]
    self.a = []
    # it should now be impossible to read any of [11, 12, 13, 14, 15, 16]
    """
    c = get_contract(code)

    assert c.foo() == [1, 2, 3, 4
```

```
>       assert c.foo() == [1, 2, 3, 4]
E       assert [11, 12, 3, 4] == [1, 2, 3, 4]
E         
E         At index 0 diff: 11 != 1
E         Use -v to get more diff
```

File: vyper/codegen/function_definitions/common.py [hunk 3]
```diff
@@ -56,6 +57,10 @@ def set_frame_info(self, frame_info: FrameInfo) -> None:
         else:
             self.frame_info = frame_info
 
+    def set_func_ir(self, func_ir: "InternalFuncIR") -> None:
```
Comment (cyberthirst) [inline vyper/codegen/function_definitions/common.py:60]: unused?

**Review by cyberthirst — APPROVED** [review]

**Comment by charles-cooper** [comment]
thanks to @trocher for the test case in https://github.com/vyperlang/vyper/pull/4037/commits/7fc2426698f1931889c105667671bf63640a4383

**Pull request merged**